contracts/guards/DelegateCallTransactionGuard.sol

```diff
@@ -0,0 +1,39 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity >=0.7.0 <0.9.0;
+
+import "../base/GuardManager.sol";
+import "../GnosisSafe.sol";
+
+contract DelegateCallTransactionGuard is Guard {
+
+    address immutable public allowedTarget;
+
+    constructor(address target) {
+        allowedTarget = target;
+    }
+
+    fallback() external {
+        // We only check known calls
+    }
+
+    function dataSeletor(bytes calldata _bytes)
+        internal
+        pure
+        returns (bytes4 selector)
+    {
+        selector =
+            _bytes[0] |
+            (bytes4(_bytes[1]) >> 8) |
+            (bytes4(_bytes[2]) >> 16) |
+            (bytes4(_bytes[3]) >> 24);
+    }
+
+    function checkCalldata(bytes calldata data, address) override view external {
+        // We only check known calls
+        if (data.length < 4 || dataSeletor(data[:4]) != GnosisSafe.execTransaction.selector) return;
+        address to;
+        uint8 operation;
+        (to,,,operation,,,,,,) = abi.decode(data[4:], (address,uint256,bytes,uint8,int256,uint256,uint256,address,address,bytes));
```

Comment:
I guess you do it this way (decoding data instead of sending all the parameters) because it's more optimal, right?

Reply:
Yes this is quite more efficient on the safe site. And also more flexible, as the fallback handler really can check the whole calldata. 

There is a drawback to it: when you upgrade the contract and the `execTransaction` method changes the `selector` changes and you need a new fallback handler. You could also say it is an advantage as it makes it very explicit what you are checking.

I will check how big the gas cost implications are and add it here.

Reply:
Execution costs changes are minimal:
From: 
```
Used 75259 gas for >transfer< with a single owner Safe (74ms)
Used 79454 gas for >transfer< with a single owner and guard Safe (80ms)
```
To:
```
Used 75270 gas for >transfer< with a single owner Safe (76ms)
Used 79381 gas for >transfer< with a single owner and guard Safe
```

Contract code size is right below the limit:
From:
`GnosisSafe 24232 bytes (limit is 24576)`
To:
`GnosisSafe 24542 bytes (limit is 24576)`

So generally I would take this approach for now ... but there might be reasons to change this in the future (as we should leave some buffer to extend the contract. ... I will try to find some other way todo so.

@Uxio0 What do you think?

Reply:
As the guard contract will need to be updated either way if we change anything on `execTransaction` I will leave the current implementation, as it's the only one that will allow us to code guards to work with multiple version of the contracts, for example:

```
if (dataSeletor(data[:4])) == ...

else if dataSeletor(data[:4]) ...
```

If we hardcode the arguments on the function we will not be able to do that and we will need different guards for different contract versions changing `execTransaction`

Reply:
you can implement a guard that implements both versions.

Also if we want to be backwards compatible in the Safe we can just set some parameters to 0 (e.g. if we remove the refund, all the refund related params would become 0). Only if we introduce new params we would have to compromise.

The other thing is that with an explicit method the code looks a lot nicer (in the guard at least, on the safe side it doesn't matter much).